File refactorings

nittka opened the pull request:
This PR addresses #134.
* refactorings have been adapted to file system location rather than workspace project relative location
* resolution unit tests for windows and unix paths (given that the paths given as arguments are correct)
* delete of source file deletes compiled files as well (analogous to rename/move)
* I have tried out the refactorings for relative, absolute and search path includes on a windows and a linux machine.

The current implementation has some small limitations
* refactoring involving virtual folders/linked files will probably fail (e.g. renaming a linked file renames only the link, not the underlying file)
* refactorings (involving search path relative includes) moving included/included files into/out of the search path may cause problems

I guess search paths are usually used for default includes which are more or less static (so refactoring will not really happen). Virtual folders/linked files will probably hardly used. So these limitations are not too problematic.

What do you think of the following enhancements for the refactoring preference page; add preferecence
* for disabling Refactoring participant completely (default refactoring enabled)
* for disabling automatic include modification (default include modification enabled)
* for disabling "delete included file" warning (default warning enabled)
* for disabling deleting compiled files when deleting a source (default deleting compiled enabled)

I will address the following issues next
* Score may produce more than one midi file, refactorings should deal with all of them
* detect virtual folders/linked files and don't do refactorings then

thSoft commented:
Thanks, I ask for your patience, I'm going to review this and your other PRs this Thursday.

thSoft commented on org.elysium.parent/org.elysium.ui/src/org/elysium/ui/refactoring/LilyPondRefactoredImportUriCalculator.java:
```diff
@@ -55,6 +69,8 @@ private String handelSearchPath(LilyPondImportUri importUri) {
 		}
 	}
 
+	//TODO ggf. kann relative Navigation vereinfacht werden
```
Please translate this comment.

thSoft commented on org.elysium.parent/org.elysium.ui/src/org/elysium/ui/refactoring/LilyPondRefactoring.java:
```diff
@@ -53,12 +57,17 @@ public LilyPondRefactoring(Operation operation, LilyPondRefactoringInjects suppo
 			checkArgumentType(arguments);
 			URI uri=URI.createPlatformResourceURI(file.getFullPath().toString(), true);
 			platformURItoFileOfRefactorTargets.put(uri, file);
+			URI fileURI = org.eclipse.emf.common.util.URI.createFileURI(file.getLocation().toFile().getAbsolutePath());
```
Please eliminate this FQN.

thSoft commented on org.elysium.parent/org.elysium.ui/src/org/elysium/ui/refactoring/LilyPondRefactoring.java:
```diff
@@ -191,16 +229,38 @@ public Change perform(IProgressMonitor pm) throws CoreException {
+	private boolean isCompiledFrom(IFile source, IFile compiled) {
+		if(source.getFullPath().removeFileExtension().equals(compiled.getFullPath().removeFileExtension())) {
+			return true;
+		} else if(LilyPondConstants.AUDIO_EXTENSION.equals(compiled.getFileExtension()) && 
```
Sorry for not mentioning it earlier, but one source file can be compiled not only to multiple audio files, but also to multiple score files. See http://lilypond.org/doc/v2.19/Documentation/notation/multiple-output-files-from-one-input-file.

nittka replied:
Could we handle this case in a separate issue? When changing my local example to this use case, I noticed that in this case the numbering pattern for the midi files goes one level deeper: score-1.pdf has score-1-1.midi and score-1-2.midi...

thSoft commented on org.elysium.parent/org.elysium.ui/src/org/elysium/ui/refactoring/LilyPondRefactoring.java:
```diff
@@ -106,13 +115,30 @@ public Operation getOperation() {
 	}
 
-	public List<IPath> getFilePaths(Set<URI> uris){
+	public Set<URI> getRefactoredFilesPlatformURIS(Set<URI> fileURIsOfIncludes){
```
Please fix the case: getRefactoredFilesPlatformURIS -> getRefactoredFilesPlatformURIs.

nittka commented on org.elysium.parent/org.elysium.ui/src/org/elysium/ui/refactoring/LilyPondRefactoredImportUriCalculator.java:
```diff
@@ -12,19 +12,19 @@
 	private IPath source;
-	//source file parent location before refactoring
+	//source file parent location after refactoring
 	private IPath sourceDestiation;
```
If you refer to "after", it is no typo - it was wrong before. Both the including and the included file could have been refactored. "source" is the including and "target" the included file, "Destination" is the respective file after the refactoring.

thSoft replied:
Sorry for not being specific, I meant destiation -> destination.

nittka commented:
I addressed your comments. If possible, I would like to extract the multiple compiled files issue to a separate ticket (as the changes require more than just removing the file extension filtering to midi). Could you also briefly comment on the potential preference page enhancement mentioned in the initial comment? If you think them worth implementing, a separate ticket could be opened.

nittka commented:
It seems, I addressed the issues and forgot to push... I am sorry back, that I missed the typo (fixed now) ;-)

thSoft commented:
Thanks for performing the requested changes! Regarding your questions:

> If possible, I would like to extract the multiple compiled files issue to a separate ticket (as the changes require more than just removing the file extension filtering to midi). 

OK.

> What do you think of the following enhancements for the refactoring preference page; add preferecence
> for disabling Refactoring participant completely (default refactoring enabled)

This would be useful.

> for disabling automatic include modification (default include modification enabled)

What are the cases where it can corrupt things?

> for disabling "delete included file" warning (default warning enabled)

This would be useful (maybe with a "Don't ask this again" in the dialog).

> for disabling deleting compiled files when deleting a source (default deleting compiled enabled)

This would also be useful.

nittka commented:
I'll open a new ticket for the preference page enhancements. Allowing to disabling the include modifications would be a safety measure. I don't know the problem cases, but if the logic destroys the includes e.g. when moving files, users may not be too happy. The preference allows users to say: "I'll deal with adapting the includes myself."

Multiple compiled files refactoring was dealt with in the last commit of this PR.
